### ISSUE TITLE
Add support for more than two compartments

### ISSUE THREAD

**Issue by ayrtonm** [immunant/IA2-Phase2#276]
This is non-trivial because each call gate is specific to a pair of compartments. This means when compartment A and B both call `foo` in C we need separate call gates for them. Currently we insert call gates using `ld --wrap` which requires the interposed symbols to be globally unique so we'll have to rename these symbols (which can be done after compiling but before linking). See #174 for the preliminary plan.

**Comment by ayrtonm** [comment]
Looks like `add_custom_command` `PRE_LINK` might make the `objcopy --redefine-syms` step mostly straighforward.

**Comment by ayrtonm** [comment]
Also see `$<TARGET_OBJECTS:some_test_target>` for object files for each target.

**Comment by ayrtonm** [comment]
Here's a rough breakdown of the changes we need to make to implement this.

- source rewriter
    - [ ] check for functions that will need multiple call gates (i.e. in the `FnDecl` pass does a function appear in more than one set in `std::set<Function> declared_fns[MAX_PKEYS]` ignoring the set that defined it)
    - [ ] add `_from_$CALLER_PKEY` suffix to the  direct call wrapper names (i.e. those that start with `__wrap`)
    - [ ] define an `llvm::raw_fd_ostream *objcopy_redefine_syms_args[MAX_PKEYS]` analogous to `ld_args_out` and lazily create these files like `write_to_ld_file` does. Only compartments that declare any of the suffixed functions need this file 
    - [ ] For each suffixed function add a line with `$FUNC $FUNC_from_$CALLER_PKEY` to the corresponding objcopy args files
- CMake
    - [ ] add a function for defining a DSO in a separate compartment (can probably modify `define_shared_lib`)
    - [ ] add the sources for the new compartment to the rewriter invocation in `define-ia2-wrapper.cmake`
    - [ ] add a flag to `define_shared_lib`/`define_test` for if the DSO needs the objcopy file (analogous to `NEEDS_LD_WRAP`)
    - [ ] use `add_custom_command` with `PRE_LINK` to get cmake to run objcopy when the objcopy flag is set (see previous comment for getting the list of object files)
    - [ ] copy `GENERATES_LD_WRAP` logic in `define-ia2-wrapper.cmake` for the objcopy flag

**Comment by ayrtonm** [comment]
I got a rough WIP working in PR #300. There are a few points where `define_shared_lib/define_test` need to be generalized for more than 3 pkeys and avoid assumptions about which DSOs are assigned which binaries. Also the objcopy step currently only supports rewriting symbols in a DSO with a single `.o`. @rinon the build system changes weren't as bad as I expected so I can finish these up myself.

The rewriter changes are done, but could use some cleanup/variable renaming to make things clearer. Also one commit just improves the error message when a compartment isn't defined so that should be split out into a separate PR.

For testing I modified `read_config` by renaming `builtin.c` to `core.c` and putting it in compartment 3. In retrospect, since plugins mainly work via indirect calls we should probably leave `builtin.c` in compartment 1 and create a another plugin `.c` for compartment 3. We should probably also add a test specifically for the new direct call gates (i.e. those that target renamed symbols) since the `read_config` test already has a lot and my changes for direct calls don't exercise that much functionality, but I'll wait until #286 is merged to redo the tests.
